src/deferred.js

```diff
@@ -212,20 +215,26 @@ jQuery.extend( {
 										}
 									};
 
+							processQueue.push( process );
+
 							// Support: Promises/A+ section 2.3.3.3.1
 							// https://promisesaplus.com/#point-57
 							// Re-resolve promises immediately to dodge false rejection from
 							// subsequent errors
 							if ( depth ) {
-								process();
+								deferredTick();
```

Comment:
This PR does introduce some sequence jumping. Note how callbacks follow an orderly progression [with the current codebase](https://jsfiddle.net/khhyxrbd/), but the `.then`s get clustered and pulled forward [by this PR](https://jsfiddle.net/khhyxrbd/1/). That's not a problem (and is in fact the whole point), but the other part is potentially a problem: that the mere introduction of a bound Deferred (i.e., constructed by a `.then` that returned a thenable) affects **all** Deferred callbacks, pulling forward to pseudo-synchronicity callbacks that previously waited for the next event loop _in front of_ other callbacks.

I'm against this PR as-is, but in favor of the broader refactoring that was hoped for when the Deferred changes first landed, to abstract out Promises/A+ compatibility code for both `Deferred#then` and `jQuery.when`, and in particular to eliminate this synchronous call (which exists to catch bad then implementations that throw after invoking a callback) by incrementing `maxDepth` earlier in the resolution process.